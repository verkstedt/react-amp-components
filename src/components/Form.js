import React, { Component } from 'react'
import { string, oneOfType, arrayOf, node, oneOf } from 'prop-types'
import { Helmet } from 'react-helmet'

import pick from '../utils/pick'
import commonPropTypes from './commonPropTypes'

class Form extends Component {
  static propTypes = {
    ...commonPropTypes,
    target: oneOf(['_blank', '_top']),
    method: oneOf(['POST', 'GET']),
    action: string,
    'action-xhr': string,
    'custom-validation-reporting': oneOf([
      'show-first-on-submit',
      'show-all-on-submit',
      'as-you-go'
    ]),
    on: string,
    children: oneOfType([
      arrayOf(node),
      node
    ])
  }

  static defaultProps = {
    target: null,
    method: 'GET',
    action: null,
    'action-xhr': null,
    'custom-validation-reporting': null,
    on: null,
    children: []
  }

  get properties() {
    return pick(this.props, ...[
      ...Object.keys(commonPropTypes),
      'target',
      'method',
      'action',
      'action-xhr',
      'custom-validation-reporting',
      'on'
    ])
  }

  render() {
    const { children } = this.props
    return [
      <Helmet key="helmet">
        <script async="" custom-element="amp-form" src="https://cdn.ampproject.org/v0/amp-form-0.1.js" />
      </Helmet>,
      <form {...{ ...this.properties }} key="form">
        {children}
      </form>
    ]
  }
}

export default Form
