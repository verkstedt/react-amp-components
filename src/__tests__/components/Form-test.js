import React from 'react'
import Form from '../../components/Form'
import Helmet from '../../utils/Helmet'
import { renderComponent } from '../test-utils'

describe('Form', () => {
  it('matches the snapshot', () => {
    const form = renderComponent(
      <Form method="GET" action="https://example.com">
        <input name="query" placeholder="search" />
        <input type="submit" value="Search" />
      </Form>
    )
    expect(form.toJSON()).toMatchSnapshot()
  })

  it('injects the right script tag', () => {
    renderComponent(
      <Form method="GET" action="https://example.com">
        <input name="query" placeholder="search" />
        <input type="submit" value="Search" />
      </Form>
    )
    Helmet.canUseDOM = false
    const staticHead = Helmet.renderStatic()
    expect(staticHead.scriptTags).toMatchSnapshot()
  })

  it('passes down the right props', () => {
    const props = {
      method: 'POST',
      action: 'https://example.com'
    }
    const form = renderComponent(
      <Form {...props}>
        <input name="query" placeholder="search" />
        <input type="submit" value="Search" />
      </Form>
    )
    expect(form.toJSON().props).toMatchObject(props)
  })
})
