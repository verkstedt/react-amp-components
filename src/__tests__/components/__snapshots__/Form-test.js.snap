// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form injects the right script tag 1`] = `
Array [
  Object {
    "async": "",
    "custom-element": "amp-form",
    "src": "https://cdn.ampproject.org/v0/amp-form-0.1.js",
  },
]
`;

exports[`Form matches the snapshot 1`] = `
<form
  action="https://example.com"
  action-xhr={null}
  custom-validation-reporting={null}
  method="GET"
  on={null}
  target={null}
>
  <input
    name="query"
    placeholder="search"
  />
  <input
    type="submit"
    value="Search"
  />
</form>
`;
